### ISSUE TITLE
Thoughts on newick format for embedding arbitrary string data

### ISSUE THREAD

**Issue by lverweijen** [glottobank/python-newick#55]
I also implemented serializing to and from newick format in https://github.com/lverweijen/littletree

There are a few things I encountered:

- My parser is very (maybe a bit too) liberal. For example, it accepts both `A(b, c, d)` as `(b, c, d)A`. Although preorder-format is not officially allowed, it reads very natural and can maybe be considered a feature, because you can visualize a function call that way. It doesn't lead to ambiguity as far as I know.
- It seems fast enough to serialize and deserialize even big trees of height and breath of 80 000.
- It accepts NHX, but no other extensions. However, it can store node comments, which can then be read with a little post-processing. Since all attributes need to be a string, some post-processing is probably needed anyway.
- It is impossible to include [ and maybe some other symbols in comments or in properties.

The thing I dislike is not being able to use [ and ] in comments and NHX-properties and not being able to serialize inputs containing these. To me the most natural solution is to allow arguments to be quoted so `node['this [[[ is possible']` becomes possible (or maybe `node[&&HNX:"key["="value[["]`, because this is already supported for node names.
On the other hand, I'm somewhat hesitant to innovate, because there seem to be too many dialects of this format already (and your implementation already supports more dialects than mine). It just seems that none of the dialects have really solved this issue or have they?

**Comment by xrotwang** [comment]
I think a major constraint on the variation that can be supported is that often Newick trees are packaged in NEXUS files. From NEXUS, the major Newick dialects inherited the rules for comments and quoting. And these two rules basically prevent any speed improvements of the "streaming Newick" type. You basically have to read to the end before you can make sense of anything at all.

Here's what the [NEXUS paper](https://doi.org/10.1093/sysbio/46.4.590) says about nested comments:

> Comments can be nested [...]. If a program encounters an open bracket within a comment, it will expect two close brackets to close both the main comment and the nested comment.

With NEXUS we have seen the slippery slope of "more liberal" reading, and I wouldn't want to repeat that for Newick :)

**Comment by lverweijen** [comment]
I'm not too familiar with Nexus, but nested comments don't seem that hard to support and I don't think they have to prevent streaming.

I was thinking of ways to escape the brackets within comments. Maybe just replacing them by xml-sequences would work.

**Comment by xrotwang** [comment]
Have you actually encountered cases of non-matching brackets in comments in Newick?

**Comment by lverweijen** [comment]
The newick format doesn't look too bad for serialization.
It is compact, readable (not as readable as prefix notation imo) and not even that hard to parse.
I don't think I really like nested comments, but if they are in the specification, we should probably support them too and if there are other tricky cases when parsing, let me know.

Now the problem is that if non-matching brackets aren't blocked, Newick can be broken in the following way:

```python
from littletree import Node

tree = Node({"closing": "]"}, identifier='simple_node')
nwk = tree.to_newick()
print(nwk)
# 'simple_node'[&&NHX:closing=]];
tree_restored = Node.from_newick(nwk)  # This would break
```

Using New Hamspire X-format it's also not possible to serialize symbols `=` and `:` (in case you want to embed a json string)
But I think I found a practical solution by xml-escaping these characters (which is not illegal in Newick afaik):

```python
tree = Node({"closing": "]"}, identifier='simple_node')
nwk = tree.to_newick(escape_comments=True)
print(nwk)
# 'simple_node'[&&NHX:closing=&rsqb;];
tree_restored = Node.from_newick(nwk)
tree_restored.show()
# simple_node
# {'closing': ']'}
```

So `]` has been encoded to `&rbsq;`.

And now it can serialize more or less everything.

**Comment by xrotwang** [comment]
I guess we look at Newick from different angles. I'm not so much concerned with "serializing everything". My main use case for Newick is interoperability. So I want to
- parse all Newick that was written with relevant standard software,
- write Newick that can be read by relevant standard software.

This is why I asked whether you have an actual case at hand of Newick used in the wild that contained non-matching brackets in comments. I haven't seen this yet and have used my `newick` package extensively on phylogenetic data from linguistic and cultural evolution (not so much on genetic data, though).

Thus, rather than extending my `newick` package to be able to handle even more complex/uncommon Newick, I welcome rather well-specified constraints like the one I quoted above.
